---
title: "Introduction_Of_LegitXMut"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction_Of_LegitXMut}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
library(knitr)
opts_chunk$set(fig.align = "center", 
               out.width = "90%",
               fig.width = 6, fig.height = 5.5,
               dev.args=list(pointsize=10),
               par = TRUE, # needed for setting hook 
               collapse = TRUE, # collapse input & ouput code in chunks
               warning = FALSE)
```

## Introduction

LegitXMut is an R package developed to analyze and visualize mutation data directly from entry of FASTA and FASTQ files. 
It can create BAM and VCF files by aligning FASTQ files to a reference genome, update VCF chromosomal names based on a 
reference, and display mutation data in different plot styles.

To download **LegitXMut**, use the following commands:

```r
require("devtools")
devtools::install_github("ZhenghaoXiao/LegitXMut", build_vignettes = TRUE)
library("LegitXMut")
```
To list all sample functions available in the package:
``` r
ls("package:LegitXMut")
```

To list all sample datasets available in the package:
``` r
#Demo data of this package was downloaded from NCBI, as listed in the reference
#Users can access them in inst/exdata/
data(package = "LegitXMut")
```

<br>


## Applications

###Aligning FASTQ to Reference Genome
To create a BAM file, align a FASTQ file to a reference genome using the alignment_FASTQ function. 
This function supports for quality control during alignment by manipulating 
indels and maxMismatches parameters.

More information about *alignment_FASTQ* function can be found using `?`
```{r, eval = FALSE}
# Load the LegitXMut package
library(LegitXMut)

# The sample data are stored in the inst/extdata/ directory of the package
fastqPath <- system.file("extdata", "ERR12205202.fastq", package = "LegitXMut")
referencePath <- system.file("extdata", "yeast.fna", package = "LegitXMut")

# Specify the output path
outputBAM <- file.path(system.file("extdata", package = "LegitXMut"), "aligned_output.bam")

# Run alignment
mutateddemoalignment <- alignment_FASTQ(
  fastqPath = fastqPath,
  referencePath = referencePath,
  indels = 10,
  maxMismatches = 1000,
  outputBAM = outputBAM
)

# See the result
View(mutateddemoalignment)

```

###Update Chromosome Names to VCF File
The update_vcf function uses the reference FASTA file to update the chromosomal names in a VCF file.
```{r, eval = FALSE}
# Load the LegitXMut package
library(LegitXMut)

# Define paths
fastaPath <- system.file("extdata", "yeast.fna", package = "LegitXMut")
vcfPath <- system.file("extdata", "aligned_output.bam.indel.VCF", package = "LegitXMut")

# Use a writable location for the output VCF file
outputVcfPath <- file.path(tempdir(), "updated.vcf")  # Temporary directory for the output

# Update VCF chromosome names
updatedvcf <- update_vcf(fastaPath, vcfPath, outputVcfPath)

```

###Visualizing Mutation Data
Input mutation data from a VCF file, plot_vcf_mutation_data creates various types of visualizations,
including heatmaps, Manhattan plots, and rainfall plots.
```{r, eval = FALSE}
library(LegitXMut)
# Define path
vcfPath <- "C:/Users/rjay1/Desktop/BCB410/LegitXMut/inst/extdata/updated.vcf"
#Plot with heatmap
plot_vcf_mutation_data(
  vcfPath = vcfPath,
  plotType = "heatmap",
  title = "Mutation Density Heatmap",
  font_size = 12,
  xlab = "Chromosome",
  ylab = "Mutation Frequency",
  legend_position = "bottom"
)
#Plot with manhattan plot
vcfPath <- "C:/Users/rjay1/Desktop/BCB410/LegitXMut/inst/extdata/updated.vcf"
plot_vcf_mutation_data(
  vcfPath = vcfPath,
  plotType = "manhattan",
  title = "Manhattan Plot of Variants",
  ylab = "-log10(Supporting Reads)",
  xlab = "Genomic Position",
  font_size = 11,
  alpha = 0.6,
  legend_position = "top"
)
#Plot with rainfall plot
vcfPath <- "C:/Users/rjay1/Desktop/BCB410/LegitXMut/inst/extdata/updated.vcf"
#Demo data only have insertion-deletion variants of more than one nucleotides changes
#Only indel will be visualized
plot_vcf_mutation_data(
  vcfPath = vcfPath,
  plotType = "rainfall",
  title = "Rainfall Plot of Mutations",
  color_scheme = c("C>A" = "red", "C>G" = "orange", "C>T" = "green",
                   "T>A" = "yellow", "T>C" = "blue", "T>G" = "purple",
                   "indel" = "grey"),
  alpha = 0.7,
  font_size = 10,
  xlab = "Genomic Position",
  ylab = "Inter-Variant Distance (log10)",
  legend_position = "right"
)
#Figures will be generated on the right corner of the R studio
```

<br>

## Package References

- Zhenghao, X. (2024) LegitXMut: an R package developed to analyze and visualize mutation data directly from entry of FASTA and FASTQ files.
 unpublished. https://github.com/ZhenghaoXiao/LegitXMut

<br>

## Other References

1.Gu, Z. et al. “circlize Implements and Enhances Circular Visualization in R.” Bioinformatics, vol. 30, no. 19, 2014, pp. 2811–2812. doi:10.1093/bioinformatics/btu393.

2.Gu, Zuguang, Roland Eils, and Matthias Schlesner. "Complex Heatmaps Reveal Patterns and Correlations in Multidimensional Genomic Data." Bioinformatics, vol. 32, no. 18, 2016, pp. 2847–2849. https://doi.org/10.1093/bioinformatics/btw313.

3.Lawrence, Michael, et al. "Software for Computing and Annotating Genomic Ranges." PLoS Computational Biology, vol. 9, no. 8, 2013, e1003118. https://doi.org/10.1371/journal.pcbi.1003118.

4.Liao, Y., Gordon K. Smyth, and Wei Shi. “The R Package Rsubread is Easier, Faster, Cheaper and Better for Alignment and Quantification of RNA Sequencing Reads.” Nucleic Acids Research, vol. 47, no. 8, 2019, e47. doi:10.1093/nar/gkz114. Available at: https://academic.oup.com/nar/article/47/8/e47/5371636.

5.Morgan, M., P. Aboyoun, R. Gentleman, M. Lawrence, and H. Pages. “GenomicAlignments: Efficient Alignments Processing in R for NGS Data.” Bioconductor, 2019, doi:10.18129/B9.bioc.GenomicAlignments. Available at: https://bioconductor.org/packages/release/bioc/html/GenomicAlignments.html.

6.Morgan, Martin, Vincent Obenchain, James Hester, and Hervé Pagès. SummarizedExperiment: Summarized Experiment Container. R package version 1.28.0, Bioconductor, 2022, https://bioconductor.org/packages/SummarizedExperiment.

7.National Center for Biotechnology Information (NCBI). Saccharomyces cerevisiae S288C Genome Assembly (GCF_000146045.2). NCBI Datasets, https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000146045.2/. Accessed 4 Nov. 2024.

8.National Center for Biotechnology Information (NCBI). Sequence Read Archive (SRA) Run: SRR29917898. NCBI SRA, https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR29917898&display=download. Accessed 4 Nov. 2024.

9.OpenAI. ChatGPT: Assistance with R Function Development for Bioinformatics Applications, "LegitxMut". https://chat.openai.com. Accessed 5 Nov. 2024.

10.Pagès, Hervé, Patrick Aboyoun, S. DebRoy, and Michael Lawrence. GenomeInfoDb: Utilities for Manipulating Chromosome Names, Including Modifying the Names to Follow a Particular Convention. R package version 1.36.0, Bioconductor, 2023, https://bioconductor.org/packages/GenomeInfoDb.

11.R Core Team. R: A Language and Environment for Statistical Computing. R Foundation for Statistical Computing, 2024, Vienna, Austria, https://www.R-project.org/.

12.Silva, Anjali. TestingPackage. GitHub, https://github.com/anjalisilva/TestingPackage. Accessed 5 Nov. 2024.

13.Wickham, Hadley. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag, 2016, https://ggplot2.tidyverse.org.

14.Wickham, Hadley, et al. dplyr: A Grammar of Data Manipulation. R package version 1.1.2, 2023, https://CRAN.R-project.org/package=dplyr.

15.Wickham, H. stringr: Simple, Consistent Wrappers for Common String Operations. R package version 1.4.0, 2019, https://CRAN.R-project.org/package=stringr.

----

```{r}
sessionInfo()
```
